### PR TITLE
Geomap panel: Generate types

### PR DESCRIPTION
**Pull request by Clarity-89**
**What is this feature?**

Schematise panel options for Geomap panel.

**Which issue(s) does this PR fix?**:

Fixes #

**Special notes for your reviewer**:
Depends on:
https://github.com/grafana/grafana/pull/61192
https://github.com/grafana/grafana/pull/61703

### CODE REVIEW

File: kinds/dashboard/dashboard_kind.cue
Comment (Clarity-89) [inline kinds/dashboard/dashboard_kind.cue:416]: Added this to veneers for dashboards as it seems to be a generic type.

File: packages/grafana-schema/src/veneer/mudball.types.ts
Comment (Clarity-89) [inline packages/grafana-schema/src/veneer/mudball.types.ts:11]: @sdboyer this seems to work, all the types defined above will overwrite the ones from mudball and the modified types will be exported. Haven't found any issues with this approach so far and it seems like nothing else is needed to make veneers work for mudball.

File: packages/grafana-schema/src/veneer/dashboard.types.ts
Comment (Clarity-89) [inline packages/grafana-schema/src/veneer/dashboard.types.ts:57]: This seems like a good candidate to automate generation for.

Comment (sdboyer) [reply packages/grafana-schema/src/veneer/dashboard.types.ts:56]: Certainly worth considering. Defaults need a fairly significant overhaul due to this fundamental limitation grafana/cuetsy#32, so we should think about any such approach in the context of addressing that problem

That's doubly important because of an idea @joshhunt and i have been circling around - rather than having just one set of defaults that are declared with the schema itself, it may make sense to have multiple, named default sets be part of the overall schema declaration story. That's something i can imagine natively including in thema.

File: packages/grafana-schema/src/schema/mudball.gen.ts
Comment (ryantxu) [inline packages/grafana-schema/src/schema/mudball.gen.ts:593]:
```suggestion
  config?: T;
```
any way to make this generic?

File: packages/grafana-schema/src/schema/mudball.gen.ts
Comment (ryantxu) [inline packages/grafana-schema/src/schema/mudball.gen.ts:589]:
```suggestion
export interface MapLayerOptions<T = any> {
```

File: packages/grafana-schema/src/schema/mudball.gen.ts
Comment (Clarity-89) [reply packages/grafana-schema/src/schema/mudball.gen.ts:593]: This is generated file from cue, which doesn't support generics. Current approach is to add a `veneer` layer that would add generics, which [done here](https://github.com/grafana/grafana/pull/61636/files#diff-6e1bcc541cb23781472d359a26f6093e1633cf070bee2bfa18d547a4fe122919R5). So basically these types are the base, which are then extended and reexported in `mudball.types.ts`.

Comment (ryantxu) [reply packages/grafana-schema/src/schema/mudball.gen.ts:593]: nice

**Comment by ryantxu** [comment]
@Clarity-89 -- this looks great!  I pushed changes to the grafana/data package so we reexport with deprecated flags, and then changed the imports.

@drew08t -- can you test and approve if all looks good?

File: packages/grafana-schema/src/common/geo.cue
Comment (ryantxu) [inline packages/grafana-schema/src/common/geo.cue:11]:
```suggestion
		// Defines a frame MatcherConfig that may filter data for the given layer
```

File: packages/grafana-schema/src/common/geo.cue
Comment (ryantxu) [inline packages/grafana-schema/src/common/geo.cue:29]:
```suggestion
```
we can remove h3... not used

File: packages/grafana-data/src/geo/layer.ts
Comment (ryantxu) [inline packages/grafana-data/src/geo/layer.ts:25]: leviate is not happy with these exports, but not sure how to do it better

File: public/app/plugins/panel/geomap/GeomapPanel.tsx
Comment (drew08t) [inline public/app/plugins/panel/geomap/GeomapPanel.tsx:159]: Where is 'degrees' coming from?

**Comment by drew08t** [comment]
> @Clarity-89 -- this looks great! I pushed changes to the grafana/data package so we reexport with deprecated flags, and then changed the imports.
> 
> @drew08t -- can you test and approve if all looks good?

Tested locally and encountered no unexpected behavior for my existing geomap dashboards and when creating a new one. Code LGTM. The only thing I'm unsure about is scaleUnits, but I think it is minor.

File: public/app/plugins/panel/geomap/GeomapPanel.tsx
Comment (Clarity-89) [reply public/app/plugins/panel/geomap/GeomapPanel.tsx:159]: ah great catch! I was testing type checking and forgot to remove this after 🙈 Will fix now.

**Review by drew08t — APPROVED** [review]

File: packages/grafana-data/src/geo/layer.ts
Comment (mckn) [reply packages/grafana-data/src/geo/layer.ts:25]: What will end up in the definition files when we are doing this? The only "actual" breaking change seems to be the `id` on the MatcherConfig that has been removed.

**Review by mckn — COMMENTED** [review]
LGTM, would be nice to see a diff on the definition files to understand why levitate is complaining. Might be something that wee need to add support for.

**Comment by Clarity-89** [comment]
> LGTM, would be nice to see a diff on the definition files to understand why levitate is complaining. Might be something that wee need to add support for.

Checked the generated `index.d.ts` file locally and these are the relevant changes: 
Before: 
```
interface MatcherConfig {
    id: string;
    options?: unknown;
}
```
After: 
```
interface MatcherConfig$1 {
    id: string;
    options?: unknown;
}

interface MatcherConfig<TConfig = any> extends MatcherConfig$1 {
    options?: TConfig;
}
```

It looks like after adding this [type extension](https://github.com/grafana/grafana/pull/61636/files#diff-932ad9420b4f3774837637528ee5bfe0ac3581cca3c5fdad61beb944aedbc45aR46) `MatcherConfig` was renamed to `MatcherConfig$1` and the newly declared `MatcherConfig` doesn't have `id` field explicitly declared. Doesn't look like a breaking change to me but more a limitation of Levitate to properly process these types of changes.

File: packages/grafana-data/src/geo/layer.ts
Comment (Clarity-89) [reply packages/grafana-data/src/geo/layer.ts:25]: Can we do a direct re-export instead? `export type { FrameGeometrySource } from '@grafana/schema';` Tried locally and seems to work, except for `FrameGeometrySourceMode` because it's used as a value, but that one is already re-exported as a value.

**Comment by ryantxu** [comment]
> Doesn't look like a breaking change to me but more a limitation of Levitate to properly process these types of changes.

@Clarity-89 -- i'll press the merge button!  nice work

**Pull request merged**
